codelabs/codelab-02/codelab2README.md

```diff
@@ -0,0 +1,501 @@
+
+```
+$ aws s3 ls s3://cmsc389l-colink
+PRE dir1/
+PRE dir2/
+2017-09-13 11 (tel:2017091311):44:17 13 hello_world.txt
+2017-09-13 12 (tel:2017091312):24:14 25165824 output.dat
+```
+
+It provides a number of operations:
+
+```
+cp ls mb mv presign rb rm sync website
+```
+
+I recommend that you play around with this tool a bit more to see how it compares with `s3api`. As you will see, it is much more limited than `s3api`, but it is quite useful for when you need to move files between your local computer and an S3 bucket.
+
+### Assignment
+
+Your assignment for this codelab is to re-create part of the `s3 sync` command using boto3 in Python. You'll be able to use this script to quickly copy files and directories into S3.
+
+**Note**: You must use boto3, not the `s3 sync` command. You will not get full credit if you just write a wrapper on top of the `sync` command.
+
+As an example of how this script works, say you had a directory like so:
+
+```
+a.txt
+folder1/
+b.txt
+c.txt
+folder2/
+folder3/
+d.txt
+```
+
+Then here are a few examples of the resulting state in S3:
+
+##### Example 1
+
+```
+$ python upload.py a.txt —bucket cmsc389l-colink
+```
+
+```
+a.txt
+```
+
+##### Example 2
+
+```
+$ python upload.py a.txt —bucket cmsc389l-colink —destination z.txt
+```
+
+```
+z.txt
+```
+
+##### Example 3
+
+```
+$ python upload.py . —bucket cmsc389l-colink
+```
+
+```
+a.txt
+folder1/b.txt
+folder1/c.txt
+folder2/folder3/d.txt
+```
+
+##### Example 4
+
+```
+$ python upload.py folder1 —bucket cmsc389l-colink —destination uploads
+```
+
+```
+uploads/b.txt
+uploads/c.txt
+```
+
+The `upload.py` file in the `code/` directory includes some boilerplate to get you started.
+
+Like the previous codelab, you only need to pass the public tests to get full credit.
+
+The tests can be run like so:
+
+```
+$ python test.py
+```
+
+Note that if you kill the testing program, it won't clean up after itself. If you end up with extra testing buckets, you can quickly remove them with the following command:
+
+```
+$ aws s3 ls | grep "test-bucket" | cut -f3 -d ' ' | xargs -I {} aws s3 rb 's3://{}'
+```
+
+Feel free to install any packages you would like to use. Just make sure to add them to the `requirements.txt` file (run `pip freeze > requirements.txt`).
+
+#### Host a Static Website
+
+As mentioned in class, you can use S3 to host static content, like static websites. Let's do that for the STICs website!
+
+Now that you've finished the assignment, you have a tool that you could use to upload static content for a website, however the tool does not set the Content-Type of each file when they are uploaded (they default to a byte stream type). If you tried to open an HTML file in your browser that was hosted on S3 with this Content-Type, then your browser would just download the file instead of serving it (it doesn't use the file ending to determine the type of file, as you might expect, it instead uses the Content-Type header in the response).
```

Comment:
Add **Important** or something else to emphasize that they need to read this section